src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
                 Token lParen = NextInvokeMemberToken();
+
                 if (lParen != null)
                 {
+                    // Fall through to the last available token expected in a member name for calculating the expected
+                    // position of the lParen. This accounts for things like `[Array]::Empty[int[]()` so that we don't
+                    // get unhandled behaviour when input is expected but incomplete.
+                    int endColumnNumber = rBracket?.Extent?.EndColumnNumber
+                        ?? genericTypeArguments?.LastOrDefault()?.Extent?.EndColumnNumber
+                        ?? member.Extent.EndColumnNumber;
+
                     Diagnostics.Assert(lParen.Kind == TokenKind.LParen || lParen.Kind == TokenKind.LCurly, "token kind incorrect");
-                    Diagnostics.Assert(member.Extent.EndColumnNumber == lParen.Extent.StartColumnNumber,
-                                       "member and paren must be adjacent");
-                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member);
+                    Diagnostics.Assert(
+                        endColumnNumber == lParen.Extent.StartColumnNumber,
+                        "member and paren must be adjacent when the method is not generic");
+                    return MemberInvokeRule(targetExpr, lParen, operatorToken, member, genericTypeArguments);
                 }
             }
 
             return new MemberExpressionAst(
-                    ExtentOf(targetExpr, member),
-                    targetExpr,
-                    member,
-                    @static: operatorToken.Kind == TokenKind.ColonColon,
-                    nullConditional: operatorToken.Kind == TokenKind.QuestionDot);
+                ExtentOf(targetExpr, member),
+                targetExpr,
+                member,
+                @static: operatorToken.Kind == TokenKind.ColonColon,
+                nullConditional: operatorToken.Kind == TokenKind.QuestionDot,
+                genericTypeArguments);
         }
 
-        private ExpressionAst MemberInvokeRule(ExpressionAst targetExpr, Token lBracket, Token operatorToken, CommandElementAst member)
+        private List<ITypeName> GenericMethodArgumentsRule(out Token rBracketToken)
+        {
+            List<ITypeName> genericTypes = null;
+
+            int resyncIndex = _tokenizer.GetRestorePoint();
+            Token lBracket = NextToken();
+            rBracketToken = null;
+
+            if (lBracket.Kind != TokenKind.LBracket)
+            {
+                // We cannot avoid this Resync(); if we use PeekToken() to try to avoid a Resync(), the method called
+                // after this (NextInvokeMemberToken()) will note that an _ungotToken is present and assume an error
+                // state. That will cause any property accesses or non-generic method calls to throw a parse error.
+                Resync(resyncIndex);
```

Comment:
If I understand correctly it should be OK to just return null here and allow the `NextInvokeMemberToken()` method to error? Or should we explicitly emit an error from here in this case?

Reply:
And do you think we can/should avoid the later resync in this method as well?

Reply:
Sorry, I was wrong about the above suggestion. We indeed cannot avoid `resync()` because `NextMemberAccessToken()` also assumes `_ungotToken == null`.